docs/import/add_disaster.js

```diff
@@ -60,14 +63,152 @@ function enableWhenReady() {
+ * reordered. Then write to firestore.
+ * @param {Object} ui
+ * @return {Promise<void>} Returns when the write to firestore finishes.
+ */
+function updateAfterSort(ui) {
+  const layerArray = disasterData.get(currentDisaster)['layerArray'];
+  const numLayers = layerArray.length;
+  const oldRealIndex = $(ui.item).children('.index-td').html();
+  const newRealIndex = numLayers - 1 - $(ui.item).index();
+
+  // pull out moved row and shuffle everything else down
+  const row = layerArray.splice(oldRealIndex, 1)[0];
+  // insert at new index
+  layerArray.splice(newRealIndex, 0, row);
+
+  // dumbly renumber all the layers. note: nth-child starts at 1, .index()
+  // starts at 0
+  for (let i = 1; i <= numLayers; i++) {
+    $('#tbody tr:nth-child(' + i + ') .index-td').html(numLayers - i);
+  }
+
+  return writeDataToFirestore();
+}
+
+/**
+ * Utility method for creating table divs with a given inner text
+ * @param {string} html inner text
+ * @return {JQuery<HTMLTableDataCellElement>}
+ */
+function createTd(html) {
+  return $(document.createElement('td')).html(html);
+}
+
+/**
+ * Create an instance of the color boxes for the color col.
+ * @param {string} color what color to make the box.
+ * @return {JQuery<HTMLDivElement>}
+ */
+function createColorBox(color) {
+  return $(document.createElement('div'))
+      .addClass('box')
+      .css('background-color', color);
+}
+
 /**
  * On change method for disaster picker.
  * @param {String} disaster
  * @return {Promise<void>} completes when we've finished filling all state
  * pickers and pulled from firebase.
  */
 function toggleDisaster(disaster) {
-  const states = disasters.get(disaster);
+  currentDisaster = disaster;
+  const data = disasterData.get(currentDisaster);
+
+  // display layer table
+  const layers = data['layerArray'];
+  const tableBody = $('#tbody');
+  tableBody.empty();
+  for (let i = layers.length - 1; i >= 0; i--) {
+    const row = $(document.createElement('tr'));
+    const layer = layers[i];
+
+    // index
+    row.append(createTd(i).addClass('index-td'));
+
+    // display name
+    // TODO: make this editable.
+    row.append(createTd(layer['display-name']));
+
+    // asset path/url sample
+    // TODO: consolidate kml-urls and urls when we have a consistent naming
+    // convention.
+    if (layer['ee-name']) {
+      row.append(createTd(layer['ee-name']));
+    } else if (layer['urls']) {
+      row.append(createTd(layer['urls'][0]));
+    } else if (layer['kml-urls']) {
+      row.append(createTd(layer['kml-urls'][0]));
+    } else if (layer['tile-urls']) {
+      row.append(createTd(layer['tile-urls'][0]));
+    }
+
+    // type
+    row.append(createTd(layerTypeStrings.get((layer['asset-type']))));
+
+    // display on load
+    const displayCheckbox =
+        $(document.createElement('input'))
+            .prop('type', 'checkbox')
+            .prop('checked', layer['display-on-load'])
+            .on('change', () => {
```

Comment:
It feels to me like this manual syncing of form<->data<->Firestore is going to get pretty repetitive. My suspicion is that something like AngularJS might be what the cool kids would do for a form that has to stay synced to a backend. Angular's blurb says in part: "Data-binding is an automatic way of updating the view whenever the model changes, as well as updating the model whenever the view changes. This is awesome because it eliminates DOM manipulation from the list of things you have to worry about."

I can imagine that we won't have time to learn/integrate Angular before the run-through, and so maybe manual coding of change handlers is the way to go, at least for now. What do you think?

If we do code change handlers ourselves, it sounds like jquery can attach a change handler to every element of a given class, so I think we should try to do that. We can then manually read the data from the row of the given changed element, which will be a little bit of unnecessary reading, and then write the whole array back to Firestore. Something like:

$('input.firestore_editable').change(function () {
   // Here, $(this) refers to the specific element of the .firestore_editable class which 'changed'
   updateLayerDataAtIndex(getIndexOfElement($(this)));
   return writeDataToFirestore();
}):

I'd recommend doing in this PR, so we know that it will work properly, because if that won't work, I would want to know alternatives to coding up 10 different change handlers before we proceed with this approach. But I'll leave up to you.

Reply:
Hmm, I think my suggestion isn't great because the entries for each row will vary based on data type. So it might be easier to trigger just the relevant data mutation, since the input being modified will have a good idea of what data needs to change. Maybe at least we can extract the JavaScript handlers to standalone functions that are all next to each other, so it's easier to keep track of them.